src/precice/tests/SerialTests.cpp

```diff
@@ -547,6 +547,69 @@ BOOST_AUTO_TEST_CASE(testExplicitWithDataInitialization)
   }
 }
 
+/**
+ * @brief Tests the reading and writing of data multiple times within one timestep.
+ *
+ * The first solver performs multiple consistent readings of data sent by the second solver.
+ * The second solver performs multiple sendings, of which the last is expected by the first solver.
+ */
+BOOST_AUTO_TEST_CASE(testExplicitWithDataMultipleReadWrite)
+{
+  PRECICE_TEST("SolverOne"_on(1_rank), "SolverTwo"_on(1_rank));
+
+  using Eigen::Vector3d;
+
+  SolverInterface cplInterface(context.name, _pathToTests + "explicit-data-multiple-read-write.xml", 0, 1);
+  if (context.isNamed("SolverOne")) {
+    int meshOneID = cplInterface.getMeshID("MeshOne");
+    cplInterface.setMeshVertex(meshOneID, Vector3d(1.0, 1.0, 1.0).data());
+    double maxDt     = cplInterface.initialize();
+    int    dataID    = cplInterface.getDataID("DataOne", meshOneID);
+    double valueData = 0.0;
+    cplInterface.initializeData();
+    cplInterface.readScalarData(dataID, 0, valueData);
+    // expected data value received
+    BOOST_TEST(3.0 == valueData);
+    // change value at read destination
+    valueData = -0.2;
+    BOOST_TEST(-0.2 == valueData);
+    cplInterface.readScalarData(dataID, 0, valueData);
+    // expected data value received
+    BOOST_TEST(3.0 == valueData);
+    while (cplInterface.isCouplingOngoing()) {
+      maxDt = cplInterface.advance(maxDt);
+      cplInterface.readScalarData(dataID, 0, valueData);
+      // expected data value received
+      BOOST_TEST(5.0 == valueData);
+      // change value at read destination
+      valueData = -0.4;
+      BOOST_TEST(-0.4 == valueData);
+      cplInterface.readScalarData(dataID, 0, valueData);
+      // expected data value received
+      BOOST_TEST(5.0 == valueData);
+    }
+    cplInterface.finalize();
+  } else {
+    BOOST_TEST(context.isNamed("SolverTwo"));
+    int meshTwoID = cplInterface.getMeshID("MeshTwo");
+    cplInterface.setMeshVertex(meshTwoID, Vector3d(0.0, 0.0, 0.0).data());
+    double maxDt  = cplInterface.initialize();
+    int    dataID = cplInterface.getDataID("DataOne", meshTwoID);
+    cplInterface.writeScalarData(dataID, 0, -0.1);
```

Comment:
```suggestion
    int size = 1;
    Eigen::VectorXi vertexIDs(size);
    Eigen::VectorXd writeData(size);
    Eigen::VectorXd writePositions(size * 3);
    vertexIDs[0] = cplInterface.setMeshVertex(meshTwoID, writePositions.data());
    double maxDt  = cplInterface.initialize();
    int    dataID = cplInterface.getDataID("DataOne", meshTwoID);
    cplInterface.writeBlockScalarData(dataID, 1, vertexIDs.data(), writeData.data())
    cplInterface.writeScalarData(dataID, vertexIDs[0], writeData[0]);
```

This is a rough draft on how to check also the other write/read functions.

Please also use `Eigen::VectorXd` and `Eigen::VectorXi`. This makes the test easier to read and allows you to directly use data.

Reply:
Do you mean to say that it's sufficient to replace one of the two `writeScalarData` with `writeBlockScalarData`?
i.e. we would be testing the writing (multiple times) of scalar data using a combination of regular and block, and not necessarily each function if that makes sense.

+ Similarly for reading and for vector data.

Reply:
Important thing here is that the functions are actually used. If in the end both functions are accessing the same data, this does not really matter in my opinion. You can use the block functions (`read/writeBlockScalar/VectorData`) for obtaining a single piece of data. Even though this is not the intended use, it should still be sufficient for this test.

Reply:
I've pushed new changes to reflect that + ready for review if current tests are sufficient.